### ISSUE TITLE
Pending improvements

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs4anki#264]
The purpose of this issue is to collect all the **non-urgent** enhancement ideas that are **not related to improving the accuracy** of the algorithm. These ideas can be implemented as and when you find it suitable.

The ideas:

- For the sake of consistency, it is suggested to either replace $w_9$ in the stability_after_lapse formula with $e^{w_9}$ or replace $e^{w_6}$ in the stability_after_recall formula with $w_6$. My suggestion would be to **replace $e^{w_6}$ in the stability_after_recall formula with $w_6$** because that is clearer.
- For the sake of clarity, it is suggested that all formulas should be re-written in such a way that all the parameters are always positive. For example, if you have D^w (w is a parameter) and w is always positive - good, don't change anything. If w is always negative - rewrite the formula as D^-w and keep w positive. This would make it easier to read the wiki.
- @Expertium suggested to rename "Average actual retention" to "Measured probability of recall" and "Optimal average actual retention" to "Theoretical optimum".
My suggestion is to
    - rename "Average actual retention" to either "Model" or "Actual Calibration"; and
    - rename "Optimal average actual retention" to "Perfect Calibration".
I would suggest renaming "number of predictions" to something like "number of items in each bin".
- To avoid confusion, it was suggested to not call R [= 0.9^(t/S) or whatever) as retention. The suggested alternatives were "probability of recall" or "retrievability". The reason for this suggestion was that the word retention has a slightly different meaning in the FSRS Stats.
If we do distinguish these terms, we should also rename `requestRetention` to `requestRetrievability`.

**Comment by Expertium** [comment]
I'll add:
1. Scheduler: remove easyBonus from the scheduler. I'm sure the new algorithm is accurate enough to work just fine without it, and it was an artifact of the old Anki algorithm anyway.
2. Add-on: change how average retention in Stats is calculated.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/dcf79263-7e5a-4350-a4c2-05fe9fc06766)

Right now it only requires t and S, so it doesn't need to parse parameters for each deck. But with the new formula for R, it will have to use deck-specific parameters. If it's not possible, we will have to remove this stat, unfortunately.
3. Optimizer: clamp parameters for grades between 0 and 10, not 1 and 10. Difficulty itself should be clamped between 1 and 10 though. During my testing I found that this benefits performance.
```
            w[14] = w[14].clamp(0, 10)
            w[15] = w[15].clamp(0, 10)
            w[16] = w[16].clamp(0, 10)
            w[17] = w[17].clamp(0, 10)
```
Also clamp initial S to prevent weird errors:
```
        if torch.equal(s, self.zero):
            # first learn, init memory states
            new_s = self.w[0] + self.w[1] * (self.w[13+x[1].long()] - self.w[14])
            new_s = new_s.clamp(0.1, 365)
```
4. Optimizer: separate the dataset into training and test sets to compare both loss and RMSE/MAE.
https://github.com/open-spaced-repetition/fsrs4anki/issues/259#issue-1714039335

**Comment by user1823** [comment]
@Expertium, the first three points in your comment are related to the experimental version of optimizer, which is not guaranteed to be released as a stable one. I created this issue to collect those ideas which we should be implemented regardless of whether we implement any of the experiments or not.

The fourth point has a separate issue for itself. So, there is no point in including it here and just adding to the distractions.

**Comment by Expertium** [comment]
Ah, ok. I'll delete my comment then.

**Comment by Expertium** [comment]
Idea for the add-on stats: distribution of difficulty.
Like this:
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/7319276c-7293-43d9-8285-353f7c52a6b6)

**Comment by L-M-Sherlock** [comment]
```python
B_W_Metric_raw.groupby(by=['d_bin']).agg('count').reset_index().plot.bar(x='d_bin', y='p', legend=False)
```

<img width="909" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/8f89c46c-eb94-4310-ae19-919879b672a6">

**Comment by user1823** [comment]
> ```python
> B_W_Metric_raw.groupby(by=['d_bin']).agg('count').reset_index().plot.bar(x='d_bin', y='p', legend=False)
> ```

The idea was for the add-on. The optimizer already has something similar.

**Comment by L-M-Sherlock** [comment]
Drawing stats graph in the add-on is unfamiliar to me. Maybe I will do some research here if I am available.

**Comment by L-M-Sherlock** [comment]
> * For the sake of clarity, it is suggested that all formulas should be re-written in such a way that all the parameters are always positive. For example, if you have D^w (w is a parameter) and w is always positive - good, don't change anything. If w is always negative - rewrite the formula as D^-w and keep w positive. This would make it easier to read the wiki.

It has been implemented in the new baseline.

**Comment by Expertium** [comment]
I'll copy this here:
I have another idea for a useful stat for the optimizer - RMSE for when y=0 ("Again") and for when y=1 ("Hard, "Good", and "Easy"). The reason is that I want to see whether RMSE for failed reviews is greater than for successful reviews, I have a suspicion that it is.
Note that bins aren't needed.
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/74124180-7171-427b-a6ca-1fe9b3230a59)
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/caacac29-fd9c-4dbf-baaa-1dea39546dab)

**Comment by L-M-Sherlock** [comment]
My current priority is to implement cross validator. I plan to use this: https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RepeatedStratifiedKFold.html#sklearn.model_selection.RepeatedStratifiedKFold

**Comment by Expertium** [comment]
That's interesting, although repeated cross-validation will increase the optimization time a lot. For example, with 3 folds and 5 repeats, optimizer will have to train 3*5=15 models.
Also, I'll copy another suggestion here:

Add the following code to the calibration graph, between plt.figure() and plt.plot(bin_prediction_means, bin_correct_means, label='Average actual retention'):
```
    ax = plt.gca()
    ax.set_xlim([0, 1])
    ax.set_ylim([0, 1])
    plt.grid(True)
```
![239749867-c47c129f-388d-4ad6-bed8-c213ab64552c](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/47395ca4-c69d-4cb0-9f5f-a4d66c00d42e)

**Comment by L-M-Sherlock** [comment]
<img width="719" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/5388b685-ff7f-4368-baaa-f25e608ac275">

I have implemented the cross validation in the new baseline: https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/baseline.ipynb

**Comment by Expertium** [comment]
Nice! Would be great if you added RMSE for each pass well, since it's more "human interpretable" than log-loss.

**Comment by L-M-Sherlock** [comment]
The current implementation of RMSE has some problems. For example, its value is affected by the number of bins.

**Comment by Expertium** [comment]
But shouldn't it be possible to calculate RMSE using every single review?
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/fb6e093c-8a65-4358-9bb7-7606ade325d8)
n is the number of reviews

**Comment by L-M-Sherlock** [comment]
> But shouldn't it be possible to calculate RMSE using every single review?

Then it will be not "human interpretable". Because the `y` for every single review is binary.

**Comment by Expertium** [comment]
Well, then we'll have to resort to using bins.

**Comment by Expertium** [comment]
@L-M-Sherlock currently, the final parameters are calculated as an arithemtic average of the parameters of all models. I think that instead it should be a weighted average, where weight is 1/test error. In other words, if the test error for that set of parameters is small, this set of parameters will have a large weight.

**Comment by L-M-Sherlock** [comment]
This test error depends on the split. Why should it be a weighted average?

**Comment by Expertium** [comment]
Maybe I'm wrong, but if a certain set of parameters leads to a lower test error, it means that this set performs better (compared to some other sets), and hence it should have more impact on how we calculate the final set (that the user will actually use). Why _shouldn't_ it be a weighted average? The arithmetic average assigns the same weight to all sets of parameters, regardless of whether the test error was low or high.

**Comment by L-M-Sherlock** [comment]
For example, we split the entire dataset into A and B, two subsets. Firstly, we train the model in A and test it in B. Secondly, we train the model in B and test it in A. We will get four errors here: A train error, B test error, B train error and A test error.

**Comment by Expertium** [comment]
> For example, we split the entire dataset into A and B, two subsets. Firstly, we train the model in A and test it in B. Secondly, we train the model in B and test it in A. We will get four errors here: A train error, B test error, B train error and A test error.

So suppose that B test error > A test error. Do you think that we should ignore this when calculating final parameters? Because I think that models should contribute based on their performance, hence weight=1/(test error).

**Comment by L-M-Sherlock** [comment]
> > For example, we split the entire dataset into A and B, two subsets. Firstly, we train the model in A and test it in B. Secondly, we train the model in B and test it in A. We will get four errors here: A train error, B test error, B train error and A test error.
> 
> So suppose that B test error > A test error. Do you think that we should ignore this when calculating the final parameters? Because I think that models should contribute based on their performance, hence weight=1/(test error). In this case, the model that was trained on B and tested on A should contribute more due to having a lower test error.

But why should we ignore the difference between the B train error and the A test error?

**Comment by user1823** [comment]
@L-M-Sherlock , the two distributions of difficulty shown by the optimizer are not equivalent. Why is it so?

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/7fd83f82-6299-4414-82fd-f964ea7a3d29)

![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/92206575/c2132c3d-2e97-4ace-b990-7efaa05e9a5d)

**Comment by Expertium** [comment]
> By the way, training phase of the optimization with the new optimizer took 4-5 times longer than the original optimizer

That is to be expected, since it effectively optimizes 3 different models.

**Comment by Expertium** [comment]
Btw, once Sherlock implements the new version with new parameters, new charts, etc., he should update the wiki entries for the [formulas](https://github.com/open-spaced-repetition/fsrs4anki/wiki/Free-Spaced-Repetition-Scheduler) and maybe for the [usage](https://github.com/open-spaced-repetition/fsrs4anki/wiki/Usage) as well. We could really use someone who is a native English speaker for this.

**Comment by L-M-Sherlock** [comment]
> the two distributions of difficulty shown by the optimizer are not equivalent. Why is it so?

Thanks for report. I find the inconsistency. I applied int(round(x)) to the previous difficulty, but applied int(x) to the current difficulty. I will fix it.

**Comment by L-M-Sherlock** [comment]
> @L-M-Sherlock , in this version, the stability calibration graph is distorted.

Thanks for your report. I'm fixing it.

**Comment by L-M-Sherlock** [comment]
I have implemented the FSRS4Anki Optimizer 4.0 Alpha version! It combines `cross-validation`, `mini-batch`, `power-difficulty`, `fixed-power-forgetting-curve`, `post-lapse-stability-offset` and `adaptive-grades`.

https://github.com/open-spaced-repetition/fsrs4anki/blob/Expt/new-baseline/experiment/fsrs4anki_optimizer_alpha.ipynb

3.19.0 vs 4.0 Alpha (my collection)
loss: 0.3163 -> 0.3124
RMSE: 0.0207 -> 0.0138
R-squared: 0.8934 -> 0.9567

It is pretty good for me. Any feedback is welcome.

**Comment by Expertium** [comment]
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/202d763b-ab88-4a2d-83d4-a9650250022c)
I get the following error

**Comment by L-M-Sherlock** [comment]
> EDIT 2: this is strange. It fixed the error for one deck, but not for another deck.

Could you share the `revlog_history.tsv` of this problematic deck with me?

By the way, I will add the clamp for new_s.

**Comment by Expertium** [comment]
I changed `n_splits` from 5 to 4 and it worked.
[revlog_history.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11564661/revlog_history.zip)
On a side note, now that both optimization and calculation of loss in 4.1 have been sped up, the bottleneck (slowest part that takes the most time) is now section 1.2, specifically this:

```
# code from https://github.com/L-M-Sherlock/anki_revlog_analysis/blob/main/revlog_analysis.py
def get_feature(x):
    last_kind = None
    for idx, log in enumerate(x.itertuples()):
        if last_kind is not None and last_kind in (1, 2) and log.type == 0:
            return x.iloc[:idx]
        last_kind = log.type
        if idx == 0:
            if log.type != 0:
                return x.iloc[:idx]
            x.iloc[idx, col_idx['delta_t']] = 0
        if idx == x.shape[0] - 1:
            break
        x.iloc[idx + 1, col_idx['i']] = x.iloc[idx, col_idx['i']] + 1
        x.iloc[idx + 1, col_idx['t_history']] = f"{x.iloc[idx, col_idx['t_history']]},{x.iloc[idx, col_idx['delta_t']]}"
        x.iloc[idx + 1, col_idx['r_history']] = f"{x.iloc[idx, col_idx['r_history']]},{x.iloc[idx, col_idx['r']]}"
    return x
```
and this:

```
def cal_retention(group: pd.DataFrame) -> pd.DataFrame:
    group['retention'] = round(group['r'].map(lambda x: {1: 0, 2: 1, 3: 1, 4: 1}[x]).mean(), 4)
    group['total_cnt'] = group.shape[0]
    return group

df = df.groupby(by=['r_history', 'delta_t'], group_keys=False).progress_apply(cal_retention)
```

**Comment by L-M-Sherlock** [comment]
> revlog_history.zip

I update the code and I test it with your data in my local. It works normally.

**Comment by Expertium** [comment]
Here, I tested the same deck
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/229a4a62-a8c6-40f0-bc7c-6f3b3e1f82b6)

**Comment by Expertium** [comment]
Woz says that the interval increase should be 7 at max, meaning that t_n/t_n-1 shouldn't be mroe than 7.

**Comment by Expertium** [comment]
@L-M-Sherlock @user1823 let's settle the naming convention once and for all.
Option 1:
The metric in the Stats window should be called "retention".
![image](https://github.com/open-spaced-repetition/fsrs4anki/assets/83031600/a0b4b241-9978-4b33-ab23-f9e8989b644f)
Whereas `requestedRetention` should be called `requestedRecall`, and all parts of the code that are NOT related to "retention" in the stats window should be renamed accordingly. I wanted to use "probability of recall", but that's too long.

Option 2:
The metric in the Stats window should be called something entirely different, though I don't know what. Then the rest of the code that uses the word "retention" can be left untouched.

Additionally, I would like another metric to be added to the stats window - estimated total knowledge, which is simply `int(average_retention * count)`. It's an estimate of how many cards user knows right now.

**Comment by Expertium** [comment]
Also, this has gotten no attention from Sherlock, so I'll mention it here: https://github.com/open-spaced-repetition/fsrs4anki/issues/247

**Comment by L-M-Sherlock** [comment]
> Also, this has gotten no attention from Sherlock, so I'll mention it here: #247

I will consider it later. I'm refactoring the `disperse siblings`.

**Comment by Expertium** [comment]
Now that https://github.com/open-spaced-repetition/fsrs4anki/issues/247 is closed, I would like to once again bring attention to [the naming convention (see above)](https://github.com/open-spaced-repetition/fsrs4anki/issues/264#issuecomment-1582282066).
@L-M-Sherlock @user1823

**Comment by user1823** [comment]
> In my option, they both are retention. The main difference is, the retention in stats is average retention. The retention in the Scheduler is retention at repetitions.

I agree and I have told this before also. However, Woz uses different terms for these, i.e. he uses the term "Retrievability" for the retention at the moment of review and the term "Retention" for the average retention in the stats.

In my opinion, we can continue using the term "Retention" for both. But, I am also OK with using different terms for these.

**Comment by L-M-Sherlock** [comment]
> Memory recall is the ability to bring things back from memory. When expressed as percentage, it can also be called retention. It is one of the measures of [retrievability](https://supermemo.guru/wiki/Retrievability) (R) for a population of memories. For example, if we remember 8 items from a list of 10 memorized items, we speak of the recall/retention of 80% (R=0.8). In a population of memories with [retrievability](https://supermemo.guru/wiki/Retrievability) R, average recall is R independent of [memory stability](https://supermemo.guru/wiki/Memory_stability).
> https://supermemo.guru/wiki/Recall

According to this quote, recall is equivalent to retention.

**Comment by L-M-Sherlock** [comment]
There is nothing named forgetting index in FSRS.

**Comment by Expertium** [comment]
Forgetting index is just 1-requestedR

**Comment by L-M-Sherlock** [comment]
There are requested retention and average retention in FSRS. Requested retention is the retention at repetitions/reviews. Average retention is the retention of all cards now. I have clarified the average retention in the Interpretation of stats.

If we need to modify the name, I don't recommend modifying requestedRetention, because the helper uses it in many functions.

**Comment by Expertium** [comment]
> In this text, Woz just wants to assert that the retention `at repetitions` is not the same as the `average` retention.

I understand, I just think that it could create confusion. But if you guys think that it's fine, then let's not change anything.

**Comment by Expertium** [comment]
Add "cards" after the number, to make it more clear. And also add a description:

> Estimated total knowledge: the number of cards that the user is expected to know today; alculated as the product of average retention and count.

Also, change the description of retention a little bit:

> Retention: the average probability of recalling a card today. In most cases, it is higher than requested retention because requested retention refers to retention at the time of a review, whereas average retention is calculated based on all cards, including undue cards.

**Comment by L-M-Sherlock** [comment]
<img width="700" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/69e262ca-da5f-4f4b-8e63-cfc3e67581c6">
Patch: [fsrs4anki-helper.zip](https://github.com/open-spaced-repetition/fsrs4anki/files/11748842/fsrs4anki-helper.zip)
Commit: https://github.com/open-spaced-repetition/fsrs4anki-helper/commit/1db346dbe3146dc609471549a6207f66b1a91529

**Comment by Expertium** [comment]
Good. One more thing and then we're done with this. Ranem "Total burden" to just "Burden", and change the description.

> Burden: an estimate of the average number of cards that have to be reviewed daily (assuming review at scheduled time). Burden = 1/I1+1/I2+ ... + 1/In where In - current interval of the n-th card.

**Comment by L-M-Sherlock** [comment]
<img width="700" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki/assets/32575846/bfeda929-bcfe-4080-a862-e880c73ce586">
